### PR TITLE
Fix: Messages marked as read incorrectly for last chat

### LINKED ISSUE #2539: Messages marked as read incorrectly for last chat [Pay on 4th May]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Expected Result:
1. Send someone a message
2. Navigate back to the LHN
3. Don't select a different chat, stay on the LHN
4. Have the person send you a message back
5. Their chat row in the LHN gets bolded to indicate it's unread 
6. The row stays bolded until you tap into it

## Actual Result:
1. Send someone a message
2. Navigate back to the LHN
3. Don't select a different chat, stay on the LHN
4. Have the person send you a message back
5. Notice that their chat row gets bolded momentarily 
6. Then, the chat row automatically gets marked as read, without you tapping on it

This is most likely happening because the app still thinks we are navigated to the last read chat, even though it isn't visible. This is because on some platforms the LHN and chat are visible together and on some platforms they aren't. 

## Action Performed:
Do the steps above on either the iOS or Android app. 

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?
No. 

## Platform:
Where is this issue occurring?

 iOS
 Android
Mobile web

**Version Number: 28-0
**Notes/Photos/Videos:

https://user-images.githubusercontent.com/1371996/115781613-3a0cf080-a380-11eb-96d1-ace25e443f3a.mp4

**Expensify/Expensify Issue URL: https://github.com/Expensify/Expensify/issues/159107

Upwork job: https://www.upwork.com/jobs/~017faab8d92bfc2427
[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

### PR DESCRIPTION
**Pull request by parasharrajat**
Please review @Julesssss 

### Details
Kindly, see the detailed proposal here https://github.com/Expensify/Expensify.cash/issues/2539#issuecomment-825181029

### Fixed Issues
<Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing>
Fixes #2539

### Tests / QA Steps

    1. Send someone a message
    2. Navigate back to the LHN
    3. Don't select a different chat, stay on the LHN
    4. Have the person send you a message back
    5. Their chat row in the LHN gets bolded to indicate it's unread
    6. The row stays bolded until you tap into it

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web / Mobile Web / Desktop

https://user-images.githubusercontent.com/24370807/115928376-a16f9100-a4a3-11eb-9bb8-cf0a2ba2e8e8.mp4

#### iOS / Android

https://user-images.githubusercontent.com/24370807/115930059-763a7100-a4a6-11eb-809c-e123e99b0776.mp4

### CODE REVIEW

File: src/pages/home/report/ReportActionsView.js [hunk 1]
```diff
@@ -26,6 +26,9 @@ import Visibility from '../../../libs/Visibility';
 import Timing from '../../../libs/actions/Timing';
 import CONST from '../../../CONST';
 import themeColors from '../../../styles/themes/default';
+import Navigation from '../../../libs/Navigation/Navigation';
+import compose from '../../../libs/compose';
+import withWindowDimensions, {windowDimensionsPropTypes} from '../../../components/withWindowDimensions';
```
Comment (Julesssss) [inline src/pages/home/report/ReportActionsView.js:31]: I'm assuming that `withWindowDimensions` is required in order to enable `isSmallScreenWidth`. is that correct?

**Review by Julesssss — DISMISSED** [review]

File: src/pages/home/report/ReportActionsView.js (hunk 1 above)
Comment (parasharrajat) [reply src/pages/home/report/ReportActionsView.js:31]: yeah

**Comment by parasharrajat** [comment]
Ready for review.

**Review by johnmlee101 — COMMENTED** [review]
I think this looks good, but I'll defer to @Julesssss for final review.

**Comment by johnmlee101** [comment]
Merge conflicts just popped up

**Comment by parasharrajat** [comment]
@johnmlee101 Just resolved it. Thanks.

**Review by Julesssss — DISMISSED** [review]

**Comment by Julesssss** [comment]
@parasharrajat a final conflict to resolve

**Comment by parasharrajat** [comment]
@Julesssss Just cleared up the Conflicts. It seems there were some revert commits so I force pushed it.

**Review by Julesssss — APPROVED** [review]
Verified again, LGTM

File: src/pages/home/report/ReportActionsView.js [hunk 2]
```diff
@@ -125,6 +140,9 @@ class ReportActionsView extends React.Component {
         // The last sequenceNumber of the same report has changed.
         const previousLastSequenceNumber = lodashGet(lastItem(prevProps.reportActions), 'sequenceNumber');
         const currentLastSequenceNumber = lodashGet(lastItem(this.props.reportActions), 'sequenceNumber');
+        const shouldRecordMaxAction = Visibility.isVisible()
+            && !(this.props.isDrawerOpen && this.props.isSmallScreenWidth);
```
Comment (marcaaron) [inline src/pages/home/report/ReportActionsView.js:144]: This logic is a bit hard to follow... can we please get rid of the `!(this && that)` logic so that this more accurately presents which scenarios we want to mark the report as read?

Comment (parasharrajat) [reply src/pages/home/report/ReportActionsView.js:144]: I am not following your comment here. This check make sure that we don't mark the actions as read when drawer is open on small screens.

On the other hand, Drawer is always open on web so we want to let it read.

Comment (marcaaron) [reply src/pages/home/report/ReportActionsView.js:144]: Might be a personal preference, but it took me some time to parse

e.g. we were to explain this to someone we might say we want to mark as read in two situations 

1. The app is visible and we are not on a small screen

OR

2. The app is visible, we are on a small screen, and the drawer is closed

Compare this to saying we want to mark as read when:

1. The app is visible AND all other situations except when the drawer is open and the screen is small.

Maybe just me, but this is easier to understand if you see it like this...

```js
// Always record on large screens. Only record on small screens when the sidebar is closed.
const shouldRecordMaxAction = Visibility.isVisible() 
    && (!this.props.isSmallScreenWidth || !this.props.isDrawerOpen)
```

lmk if it looks the same to you 😂

Comment (parasharrajat) [reply src/pages/home/report/ReportActionsView.js:144]: Yup. Will update.
